packages/skygear-react-native/android/src/main/java/io/skygear/reactnative/SGSkygearReactNativeModule.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    public String getName() {
+        return "SGSkygearReactNative";
+    }
+
+    @ReactMethod
+    public void randomBytes(int length, Promise promise) {
+        SecureRandom rng = new SecureRandom();
+        byte[] output = new byte[length];
+        rng.nextBytes(output);
+        promise.resolve(this.bytesToArray(output));
+    }
+
+    @ReactMethod
+    public void sha256String(String input, Promise promise) {
+        try {
+            MessageDigest md = MessageDigest.getInstance("SHA-256");
+            md.update(input.getBytes(Charset.forName("UTF-8")));
+            promise.resolve(this.bytesToArray(md.digest()));
+        } catch (NoSuchAlgorithmException e) {
+            promise.reject(e);
+        }
+    }
+
+    @ReactMethod
+    public void openURL(String urlString, String scheme, Promise promise) {
+        this.openURLPromise = promise;
+
+        if (this.openURLReceiver != null) {
+            this.reactContext.getApplicationContext().unregisterReceiver(this.openURLReceiver);
+            this.openURLReceiver = null;
+        }
+
+        this.openURLReceiver = new BroadcastReceiver() {
+            @Override
+            public void onReceive(Context context, Intent intent) {
+                SGSkygearReactNativeModule.this.onIntent(intent);
+            }
+        };
+
+        IntentFilter filter = new IntentFilter();
+        filter.addAction(OAUTH_ACTION);
+        filter.addDataScheme(scheme);
+        this.reactContext.getApplicationContext().registerReceiver(this.openURLReceiver, filter);
+
+        try {
+            Activity currentActivity = getCurrentActivity();
+            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(urlString).normalizeScheme());
+
+            String selfPackageName = getReactApplicationContext().getPackageName();
+            ComponentName componentName =
+                          intent.resolveActivity(getReactApplicationContext().getPackageManager());
+            String otherPackageName = (componentName != null ? componentName.getPackageName() : "");
+            // If there is no currentActivity or we are launching to a different package we need to set
+            // the FLAG_ACTIVITY_NEW_TASK flag
+            if (currentActivity == null || !selfPackageName.equals(otherPackageName)) {
+                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+            }
+
+
+            if (currentActivity != null) {
+                currentActivity.startActivity(intent);
+            } else {
+                getReactApplicationContext().startActivity(intent);
+            }
+        } catch (Exception e) {
+            if (this.openURLPromise != null) {
+                this.openURLPromise.reject(e);
+                this.openURLPromise = null;
+            }
+        }
+    }
+
+    @ReactMethod
+    public void signInWithApple(String urlString, Promise promise) {
+        promise.reject("EUNSPECIFIED", "Android does not support Sign in with Apple");
```

Comment:
I think for RN Android, we'll open chrome and use default OIDC flow?

Reply:
I wanted to make the API simpler. `loginApple` always means iOS AuthenticationServices  flow while `loginOAuthProvider` always means OIDC flow.